addon/test-waiter.ts

```diff
@@ -0,0 +1,97 @@
+import { IWaiter, WaiterName, ITestWaiterDebugInfo } from './types';
+import { register } from './waiter-manager';
+
+/**
+ * A class providing creation, registration and async waiting functionality.
+ *
+ * @public
+ * @class TestWaiter<T>
+ */
+export default class TestWaiter<T> implements IWaiter {
+  public name: WaiterName;
+  private isRegistered = false;
+
+  items = new Map<T, ITestWaiterDebugInfo>();
+
+  /**
+   * @public
+   * @constructor
+   * @param name {WaiterName} the name of the test waiter
+   */
+  constructor(name: WaiterName) {
+    this.name = name;
+  }
+
+  /**
+   * Will register the waiter, allowing it to be opted in to pausing async
+   * operations until they're completed within your tests. You should invoke
+   * it after instantiating your `TestWaiter` instance.
+   *
+   * **Note**, if you forget to register your waiter, it will be registered
+   * for you on the first invocation of `beginAsync`.
+   *
+   * @private
+   * @method register
+   */
+  private register() {
+    if (!this.isRegistered) {
+      register(this);
+      this.isRegistered = true;
+    }
+  }
+
+  /**
+   * Should be used to signal the beginning of an async operation that
+   * is to be waited for. Invocation of this method should be paired with a subsequent
+   * `endAsync` call to indicate to the waiter system that the async operation is completed.
+   *
+   * @public
+   * @method beginAsync
+   * @param item {T} The item to register for waiting
+   * @param label {string} An optional label to identify the item
+   */
+  beginAsync(item: T, label?: string) {
+    this.register();
+
+    this.items.set(item, {
+      stack: new Error().stack,
```

Comment:
Nothing to do here ATM, but I wonder if we should create the error and not forcefully/eagerly do the stack generation. In most cases we won't use the stack, and I believe generating one is somewhat costly...

Reply:
Yep, this is what we do in Backburner, so I think we do the same here.

Reply:
Done.